CSharpSourceCode/Battle/Grenades/GrenadesMissionLogic.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System.Linq;
+using TaleWorlds.Core;
+using TaleWorlds.Engine;
+using TaleWorlds.InputSystem;
+using TaleWorlds.Library;
+using TaleWorlds.MountAndBlade;
+using TaleWorlds.ObjectSystem;
+using TOW_Core.Utilities;
+
+namespace TOW_Core.Battle.Grenades
+{
+    public class GrenadesMissionLogic : MissionLogic
+    {
+        public GrenadesMissionLogic()
+        {
+        }
+
+        public override void OnMissionTick(float dt)
+        {
+            base.OnMissionTick(dt);
+            if (Input.IsKeyPressed(InputKey.L))
+                EquipGrenades();
+        }
+        public override void OnAgentShootMissile(Agent shooterAgent, EquipmentIndex weaponIndex, Vec3 position, Vec3 velocity, Mat3 orientation, bool hasRigidBody, int forcedMissileIndex)
+        {
+            base.OnAgentShootMissile(shooterAgent, weaponIndex, position, velocity, orientation, hasRigidBody, forcedMissileIndex);
+            ActivateIfGrenade();
+        }
+        public override void OnAgentHit(Agent affectedAgent, Agent affectorAgent, int damage, in MissionWeapon affectorWeapon)
+        {
+            base.OnAgentHit(affectedAgent, affectorAgent, damage, affectorWeapon);
+            if (affectorWeapon.Item != null && affectorWeapon.Item.Name.Contains("Grenade"))
+            {
+                Mission.Missile grenade = Mission.Current.Missiles.ToList().First(m => m.Weapon.Item.Name.Contains("Grenade"));
```

Comment:
This will break when there is more than one grenade in the mission.

Something like this should assure you find the correct missile not just the first one that you happen to iterate to.

> var id = affectorWeapon.Item.Id;
> Mission.Missile grenade = Mission.Current.Missiles.ToList().First(m => m.Weapon.Item.Id == id);

Im also not sure about potential performance impact of running this on every hit? 
Thoughts @hunharibo @tristanka @Z3rca ?

Reply:
Okay, i'll test with id. But about using OnAgentHit...i'm not sure that there is another way. At least i'll try to handle missile collision reaction.